rust/agents/relayer/src/msg/gelato_submitter.rs

```diff
@@ -1,98 +1,231 @@
+    /// Intended to be shared by reqwest library.
+    pub(crate) http: reqwest::Client,
+    /// Prometheus metrics.
+    /// TODO(webbhorn): Promote to non-_-prefixed name once we're populating metrics.
+    pub(crate) _metrics: GelatoSubmitterMetrics,
 }
 
 impl GelatoSubmitter {
-    pub fn new(
-        cfg: GelatoConf,
-        rx: mpsc::UnboundedReceiver<SubmitMessageArgs>,
-        inbox_contracts: InboxContracts,
-        db: AbacusDB,
-    ) -> Self {
-        assert!(cfg.enabled_for_message_submission);
-        Self {
-            rx,
-            inbox_contracts,
-            db,
-        }
-    }
-
     pub fn spawn(mut self) -> Instrumented<JoinHandle<Result<()>>> {
         tokio::spawn(async move { self.work_loop().await })
-            .instrument(info_span!("submitter work loop"))
+            .instrument(info_span!("gelato submitter work loop"))
     }
 
-    /// The Gelato relay framework allows us to submit ops in
-    /// parallel, subject to certain retry rules. Therefore all we do
-    /// here is spin forever asking for work from the rx channel, then
-    /// spawn the work to submit to gelato in a root tokio task.
-    ///
-    /// It is possible that there has not been sufficient interchain
-    /// gas deposited in the InterchainGasPaymaster account on the source
-    /// chain, so we also keep a wait queue of ops that we
-    /// periodically scan for any gas updates.
-    ///
-    /// In the future one could maybe imagine also applying a global
-    /// rate-limiter against the relevant Gelato HTTP endpoint or
-    /// something, or a max-inflight-cap on Gelato messages from
-    /// relayers, enforced here. But probably not until that proves to
-    /// be necessary.
     async fn work_loop(&mut self) -> Result<()> {
         loop {
             self.tick().await?;
-            tokio::task::yield_now().await;
+            sleep(Duration::from_millis(1000)).await;
         }
     }
 
-    /// Extracted from main loop to enable testing submitter state
-    /// after each tick, e.g. in response to a change in environment
-    /// conditions like values in InterchainGasPaymaster.
-    async fn tick(&self) -> Result<()> {
-        // TODO(webbhorn): Pull all available messages out of self.rx
-        // and check if enough gas to process them. If not, put on
-        // wait queue. If there is enough, spawn root task and run
-        // the fwd req op.
+    async fn tick(&mut self) -> Result<()> {
+        // Pull any messages sent by processor over channel.
+        loop {
+            match self.messages.try_recv() {
+                Ok(msg) => {
+                    let op = ForwardRequestOp {
+                        args: self.make_forward_request_args(msg)?,
+                        opts: ForwardRequestOptions::default(),
+                        signer: self.signer.clone(),
+                        http: self.http.clone(),
+                    };
+                    tokio::spawn(async move {
```

Comment:
Hmm, would kinda like to see more tracking rather than just letting them float freely. However that is something that we could come back to as it is mostly code addition and not any significant change to what is here now.

Reply:
I agree. In my work-in-progress branch, I track the generated work in a FuturesUnordered and then select! between that and the channel that receives incoming messages from the processor. https://github.com/abacus-network/abacus-monorepo/blob/3cae74f8e65e2d552b79ff9e78df1476a279519e/rust/agents/relayer/src/msg/gelato_submitter.rs if you want a preview. Let me know if you are ok with me leaving this here pending next PR or want me to clean it up before submitting.